[RND-543] Run Meadowlark in Containers on Azure (#245)
Ed-Fi-Exchange-OSS/Meadowlark: eng/deploy/azure/README.md

# Azure Deployment

To deploy to Azure, this can be done through Azure Container Instances (ACI),
with the [Docker Azure
Integration](https://docs.docker.com/cloud/aci-integration/), or with [Azure
CLI](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli).

## Deploy with Docker Azure Integration

- [Log into Azure from
  Docker](https://docs.docker.com/cloud/aci-integration/#log-into-azure).

- [Create an ACI Docker
  context](https://docs.docker.com/cloud/aci-integration/#create-an-aci-context)

- Browse to `../eng/deploy/azure`

- Create a .env file. Update URLs to match your correct Azure region.

- Execute the following script:

```Shell

# Switch to the ACI Context
docker context use myacicontext

# Login to Azure Container Registry
az acr login --name edfimeadowlark

docker compose -p meadowlark --file ./azure-docker-compose.yml up -d

# Initialize replica set

> az container exec --resource-group {resource group name} -n meadowlark
> --container-name ml-mongo1 --exec-command 'mongo --eval rs.initiate()'

Reviewer: Needs a line continuation character:

```suggestion
az container exec --resource-group {resource group name} -n meadowlark `
   --container-name ml-mongo1 --exec-command 'mongo --eval rs.initiate()'
```

```

## Deploy with Azure CLI

For Azure CLI, it's necessary to specify all environment variables in the
command line since it is not possible to read a .env file.

```Shell
# Login to Azure
az login

# Login to Azure Container Registry
az acr login --name edfimeadowlark

# Create the mongo container

> az container create --resource-group {resource group name} -n ml-mongo
> --image mongo:4.0.28 --ports 27017 --dns-name-label mlmongo1
> --command-line "mongod --replSet rs0"
> 
> # Initialize mongodb replica set
> az container exec --resource-group {resource group name} -n meadowlark
> --container-name ml-mongo1 --exec-command 'mongo --eval rs.initiate()'
> 
> # Create OpenSearch container
> az container create --resource-group {resource group name} -n meadowlark
> --image edfimeadowlark.azurecr.io/meadowlark-opensearch:latest --dns-name-label ml-opensearch1
> 
> # Create meadowlark container
> az container create --resource-group {resource group name} -n meadowlark
> --image edfialliance/meadowlark-ed-fi-api:pre --ports 80 --environment-variables {specify all env variables required}
> 
> 

Reviewer: ```suggestion
az container create --resource-group {resource group name} -n ml-mongo `
    --image mongo:4.0.28 --ports 27017 --dns-name-label mlmongo1 `
    --command-line "mongod --replSet rs0"

# Initialize mongodb replica set
az container exec --resource-group {resource group name} -n meadowlark `
    --container-name ml-mongo1 --exec-command 'mongo --eval rs.initiate()'

# Create OpenSearch container
az container create --resource-group {resource group name} -n meadowlark `
    --image edfimeadowlark.azurecr.io/meadowlark-opensearch:latest
    --dns-name-label ml-opensearch1

# Create meadowlark container
az container create --resource-group {resource group name} -n meadowlark `
    --image edfialliance/meadowlark-ed-fi-api:pre --ports 80 `
    --environment-variables {specify all env variables required}
```

# Azure Deployment

To deploy to Azure, this can be done through Azure Container Instances (ACI),
with the [Docker Azure
Integration](https://docs.docker.com/cloud/aci-integration/), or with [Azure
CLI](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli).

## Deploy with Docker Azure Integration

> - [Log into Azure from
>   Docker](https://docs.docker.com/cloud/aci-integration/#log-into-azure).
> 
> - [Create an ACI Docker
>   context](https://docs.docker.com/cloud/aci-integration/#create-an-aci-context)
> 
> - Switch to the ACI Context
> 
> ```Shell
> docker context use myacicontext
> ```
> 
> - Login to Azure Container Registry
> 
> ```Shell
> az acr login --name edfimeadowlark
> ```

Reviewer: I was thinking something more like this, except you would need to add the rest of the commands into this same Shell block.

```suggestion
- [Log into Azure from
  Docker](https://docs.docker.com/cloud/aci-integration/#log-into-azure).
- [Create an ACI Docker
  context](https://docs.docker.com/cloud/aci-integration/#create-an-aci-context)
- Switch to the ACI Context and Deploy Meadowlark

   ```Shell
   docker context use myacicontext
   az acr login --name edfimeadowlark
   ````
```

- [Create an ACI Docker
  context](https://docs.docker.com/cloud/aci-integration/#create-an-aci-context)

- Browse to `../eng/deploy/azure`

- Create a .env file. Update URLs to match your correct Azure region.

- Execute the following script:

```Shell

# Switch to the ACI Context
docker context use myacicontext

# Login to Azure Container Registry
az acr login --name edfimeadowlark

docker compose -p meadowlark --file ./azure-docker-compose.yml up -d

# Initialize replica set
az container exec --resource-group {resource group name} -n meadowlark

> --container-name ml-mongo1 --exec-command 'mongo --eval rs.initiate()'

Reviewer: Won't this fail without a line continuation character?
```suggestion
az container exec --resource-group {resource group name} -n meadowlark `
    --container-name ml-mongo1 --exec-command 'mongo --eval rs.initiate()'
```

Annoying thing is that I entered the PowerShell line continuation, which won't work for Bash. And vice versa. I'm OK with that though.
